File: 31_stack_sequence/main.py
# -*- coding: utf-8 -*-


def is_pop_order(push, pop):
    possible = False
    if len(push) > 0 and len(pop) > 0 and len(push) == len(pop):
        s, push_idx, pop_idx = [], 0, 0

        while pop_idx < len(pop):
            while len(s) == 0 or s[-1] != pop[pop_idx]:
                if push_idx == len(push):
                    break
                s.append(push[push_idx])
                push_idx += 1
            if s[-1] != pop[pop_idx]:
                break
            s.pop()
            pop_idx += 1

        if len(s) == 0 and pop_idx == len(pop):
            possible = True
    return possible


if __name__ == '__main__':
    print(is_pop_order([1, 2, 3, 4, 5], [5, 4, 3, 2, 1]))
    print(is_pop_order([1, 2, 3, 4, 5], [5, 3, 4, 1, 2]))

File: 31_stack_sequence/go/main.go
package main

import (
	"fmt"
)

func IsPopOrder(push, pop []int) bool {
	possible := false
	if len(push) > 0 && len(pop) > 0 && len(push) == len(pop) {
		s, pushIdx, popIdx := make([]int, 0, len(push)), 0, 0
		for popIdx < len(pop) {
			for len(s) == 0 || s[len(s) - 1] != pop[popIdx] {
				if pushIdx == len(push) {
					break
				}
				s = append(s, push[pushIdx])
				pushIdx++
			}
			if s[len(s) - 1] != pop[popIdx] {
				break
			}
			s = s[:len(s) - 1]
			popIdx++
		}
		if len(s) == 0 && popIdx == len(pop) {
			possible = true
		}
	}
	return possible
}

func main() {
	fmt.Println(IsPopOrder([]int{1, 2, 3, 4, 5}, []int{5, 4, 3, 2, 1}))
	fmt.Println(IsPopOrder([]int{1, 2, 3, 4, 5}, []int{5, 3, 4, 1, 2}))
}
